src/CodeGen_LLVM.cpp

```diff
@@ -4261,17 +4270,94 @@ void CodeGen_LLVM::visit(const Evaluate *op) {
-        vector<Value *> vecs;
-        for (Expr i : op->vectors) {
-            vecs.push_back(codegen(i));
-        }
         value = interleave_vectors(vecs);
     } else {
-        vector<Value *> vecs;
-        for (Expr i : op->vectors) {
-            vecs.push_back(codegen(i));
+        // If the even-numbered indices equal the odd-numbered
+        // indices, only generate one and then do a self-interleave.
+        for (int f : {4, 3, 2}) {
+            bool self_interleave = op->indices.size() % f == 0;
+            for (size_t i = 0; i < op->indices.size(); i++) {
+                self_interleave &= (op->indices[i] == op->indices[i - (i % f)]);
+            }
+            if (self_interleave) {
+                vector<int> sub_indices;
+                for (size_t i = 0; i < op->indices.size(); i += f) {
+                    sub_indices.push_back(op->indices[i]);
+                }
+                Expr equiv = Shuffle::make(op->vectors, sub_indices);
+                value = codegen(equiv);
+                if (f == 2) {
+                    value = interleave_vectors({value, value});
+                } else if (f == 3) {
+                    value = interleave_vectors({value, value, value});
+                } else {
+                    value = interleave_vectors({value, value});
+                    value = interleave_vectors({value, value});
+                }
+                return;
+            }
+
+            // Check for an interleave of slices (i.e. an in-vector transpose)
+            bool interleave_of_slices = op->vectors.size() == 1 && op->indices.size() % f == 0;
+            int step = op->type.lanes() / f;
+            for (int i = 0; i < step; i++) {
+                for (int j = 0; j < f; j++) {
+                    interleave_of_slices &= (op->indices[i * f + j] == j * step + i);
+                }
+            }
+            if (interleave_of_slices) {
+                value = codegen(op->vectors[0]);
+                vector<Value *> slices;
+                for (int i = 0; i < f; i++) {
+                    slices.push_back(slice_vector(value, i * step, step));
+                }
+                value = interleave_vectors(slices);
+                //optimization_barrier(value);
+            }
+        }
+        // If the indices form contiguous aligned runs, do the shuffle
+        // on entire sub-vectors by reinterpreting them as a wider
+        // type.
+        for (int f : {8, 4, 2}) {
+
+            if (op->type.lanes() % f != 0) {
+                continue;
+            }
+
+            if (op->type.bits() * f > 64) {
+                continue;
+            }
+            bool contiguous = true;
+            for (const Expr &vec : op->vectors) {
+                contiguous &= ((vec.type().lanes() % f) == 0);
+            }
+            for (size_t i = 0; i < op->indices.size(); i++) {
+                contiguous &= (op->indices[i] == op->indices[i - (i % f)] + ((int)i % f));
```

Comment:
It might be easier/more readable to just construct the shuffles you are looking for and compare for equality?

Reply:
Improved it by changing to a nested loop (and caught a bug!)